src/main.cpp

```diff
@@ -3953,8 +3984,56 @@ void static ProcessGetData(CNode* pfrom)
+                            pfrom->PushMessage("block", block);
+                            LogPrint("thin", "Sent regular block instead - thinblock size: %d vs block size: %d => tx hashes: %d transactions: %d  peerid=%d\n", nSizeThinBlock, nSizeBlock, thinBlock.vTxHashes.size(), thinBlock.mapMissingTx.size(), pfrom->id);
+                        }
+                    }
+                    else if (inv.type == MSG_XTHINBLOCK)
+                    {
+                        CXThinBlock xThinBlock(block, *pfrom->pfilter);
+                        int nSizeBlock = ::GetSerializeSize(block, SER_NETWORK, PROTOCOL_VERSION);
+                        if (xThinBlock.collision == true) // If there is a cheapHash collision in this block then send a normal thinblock
+                        {
+                            CThinBlock thinBlock(block, *pfrom->pfilter);
+                            int nSizeThinBlock = ::GetSerializeSize(xThinBlock, SER_NETWORK, PROTOCOL_VERSION);
+                            if(nSizeThinBlock < nSizeBlock) {
+                                pfrom->PushMessage("thinblock", thinBlock);
+                                LogPrint("thin", "TX HASH COLLISION: Sent thinblock - size: %d vs block size: %d => tx hashes: %d transactions: %d  peerid=%d\n", nSizeThinBlock, nSizeBlock, xThinBlock.vTxHashes.size(), xThinBlock.mapMissingTx.size(), pfrom->id);
+                            }
+                            else {
+                                pfrom->PushMessage("block", block);
+                                LogPrint("thin", "Sent regular block instead - xthinblock size: %d vs block size: %d => tx hashes: %d transactions: %d  peerid=%d\n", nSizeThinBlock, nSizeBlock, xThinBlock.vTxHashes.size(), xThinBlock.mapMissingTx.size(), pfrom->id);
+                            }
+                        }
+                        else // Send an xThinblock
+                        {
+                            // Only send a thinblock if smaller than a regular block
+                            int nSizeThinBlock = ::GetSerializeSize(xThinBlock, SER_NETWORK, PROTOCOL_VERSION);
+                            if(nSizeThinBlock < nSizeBlock) {
+                                pfrom->PushMessage("xthinblock", xThinBlock);
+                                LogPrint("thin", "Sent xthinblock - size: %d vs block size: %d => tx hashes: %d transactions: %d  peerid=%d\n", nSizeThinBlock, nSizeBlock, xThinBlock.vTxHashes.size(), xThinBlock.mapMissingTx.size(), pfrom->id);
+                            }
+                            else {
+                                pfrom->PushMessage("block", block);
+                                LogPrint("thin", "Sent regular block instead - xthinblock size: %d vs block size: %d => tx hashes: %d transactions: %d  peerid=%d\n", nSizeThinBlock, nSizeBlock, xThinBlock.vTxHashes.size(), xThinBlock.mapMissingTx.size(), pfrom->id);
+                            }
+                        }
+                    }
```

Comment:
please make this whole chunk a separate function (in a separate file)... something like "HandleReceivedThinBlock".  I know that it is a bit awkward, but rebasing is so much easier if the changes are a single function call with a well-defined interaction to the existing code.